_posts/2019-01-10-moving-client-packages-to-monorepo.md

```diff
@@ -0,0 +1,82 @@
+
+For example, we have a package called [@sensenet/client-utils](https://www.npmjs.com/package/@sensenet/client-utils) and I want to add a new helper function there.
+Let's say a debounce that I will use in another package.
+
+### The workflow is the following for this:
+
+1.  Create a branch from develop in the client-utils package
+2.  Create the debounce function
+3.  Create a pull request and wait for the code review
+4.  Merge back to develop
+5.  Do a release
+6.  Do a package upgrade in the package I want to use the util function
+7.  Notice a bug in the debounce function and start a process again
+
+I know, there shouldn't be a bug in the code that have been reviewed. Also we could just copy the built code to the package node_modules folder to use that instead of the package from npm but that felt like a pain in the long run. There must be an easier way for this.
+
+## Monorepo to the rescue
+
+> In revision control systems, a monorepo (syllabic abbreviation of monolithic repository) is a software development strategy where code for many projects are stored in the same repository.
+
+**In a monorepo the whole workflow can be simplified like this:**
+
+1.  Create a branch from develop
+2.  Create the function
+3.  Build, test
+4.  Merge back when done
+
+So much easier and cleaner than before. No need to copy anything, anywhere. No need to release a dev alpha or a minor version. You can see your changes in all the packages once you modified something. It not only made our life simpler but it gave us a productivity boost as well. Let's see how the transition went.
+
+![Patrick](/img/posts/monorepo/patrick-meme.jpg 'Get all the packages we extracted from one repository and put them together in a new one')
+
+## Moving sensenet client packages to monorepo
+
+First thing first, we need a new reopsitory. We could also use one of the repos and put everything there, but it felt cleaner to start a new one. I have created a todo list in this [pull request](https://github.com/SenseNet/sn-client/pull/1) to track the progress and to see what is left. 
+
+We moved all the packages to the packages folder with lerna. _[Lerna](https://lernajs.io/) is a tool that optimizes the workflow around managing multi-package repositories with git and npm._ This way we could see the commit history in the new repository as well. We had to make sure that every package is able to build itslef. We faced with some problems here. Not every package was configured correctly. Some was missing dependencies that were needed for it to be able to build.
+
+After that we run the tests to make sure we didn't break anything. Because the repositories were not consistent in their configs, it failed. We had to temporarily suspend the linting to see the test results. With that said we realized that we should to make all repositories consistent. Fortunately there were no big errors.
```

Comment:
```suggestion
After that we run the tests to make sure we didn't break anything. Because the repositories were not consistent in their configs, it failed. We had to temporarily suspend the linting to see the test results. With that said we realized that we should make all repositories consistent. Fortunately there were no big errors.
```